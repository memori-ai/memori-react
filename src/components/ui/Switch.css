.memori-switch-wrapper {
    display: inline-flex;
    align-items: center;
    gap: 0.5rem;
  }
  
  .memori-switch--label {
    color: #333;
    font-size: 14px;
    font-weight: 500;
  }
  
  .memori-switch {
    position: relative;
    display: inline-flex;
    width: 44px;
    height: 22px;
    align-items: center;
    border: none;
    border-radius: 100px;
    background-color: rgba(0, 0, 0, 0.25);
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.1);
    cursor: pointer;
    outline: none;
    transition: all 0.2s;
  }
  
  /* Size variants */
  .memori-switch--small {
    width: 28px;
    min-width: 28px;
    height: 16px;
  }
  
  .memori-switch--large {
    width: 56px;
    min-width: 56px;
    height: 28px;
  }
  
  .memori-switch:focus {
    box-shadow: 0 0 0 2px rgba(24, 144, 255, 0.2);
  }
  
  .memori-switch--checked {
    background-color: #1890ff;
  }
  
  .memori-switch--disabled {
    background-color: rgba(0, 0, 0, 0.25); /* Stay gray when disabled */
    cursor: not-allowed;
    opacity: 0.4;
  }
  
  .memori-switch--checked.memori-switch--disabled {
    background-color: rgba(24, 144, 255, 0.6); /* Light blue when checked and disabled */
  }
  
  .memori-switch--handle {
    position: absolute;
    top: 2px;
    left: 2px;
    display: flex;
    width: 18px;
    height: 18px;
    align-items: center;
    justify-content: center;
    border-radius: 50%;
    background-color: #fff;
    box-shadow: 0 2px 4px 0 rgba(0, 35, 11, 0.2);
    transform: translateX(0%);
    transition: all 0.2s;
  }
  
  .memori-switch--small .memori-switch--handle {
    width: 12px;
    height: 12px;
  }
  
  .memori-switch--large .memori-switch--handle {
    width: 24px;
    height: 24px;
  }
  
  .memori-switch--handle-checked {
    transform: translateX(22px);
  }
  
  .memori-switch--small .memori-switch--handle-checked {
    transform: translateX(12px);
  }
  
  .memori-switch--large .memori-switch--handle-checked {
    transform: translateX(28px);
  }
  
  .memori-switch--inner {
    margin-right: 6px;
    margin-left: 24px;
    color: #fff;
    font-size: 12px;
  }
  
  .memori-switch--checked .memori-switch--inner {
    margin-right: 24px;
    margin-left: 6px;
  }
  
  /* Loading spinner animation */
  .memori-switch--loading-spinner {
    width: 12px;
    height: 12px;
    animation: switch-spinner 1s linear infinite;
  }
  
  .memori-switch--small .memori-switch--loading-spinner {
    width: 8px;
    height: 8px;
  }
  
  .memori-switch--large .memori-switch--loading-spinner {
    width: 16px;
    height: 16px;
  }
  
  @keyframes switch-spinner {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }