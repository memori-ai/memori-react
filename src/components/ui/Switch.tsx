import React, { useState, useEffect } from 'react';
import { Switch as HeadlessSwitch } from '@headlessui/react';
import cx from 'classnames';

export interface Props {
  className?: string;
  checked?: boolean;
  defaultChecked?: boolean;
  onChange?: (checked: boolean) => void;
  disabled?: boolean;
  size?: 'small' | 'default' | 'large';
  checkedChildren?: React.ReactNode;
  unCheckedChildren?: React.ReactNode;
  loading?: boolean;
  'data-testid'?: string;
  label?: string;
  id?: string;
}

const Switch = ({
  className,
  checked,
  defaultChecked = false,
  onChange,
  disabled = false,
  size = 'default',
  checkedChildren,
  unCheckedChildren,
  loading = false,
  'data-testid': testId,
  label,
  id,
}: Props) => {
  // Handle controlled vs uncontrolled
  const [innerChecked, setInnerChecked] = useState(defaultChecked);
  const isControlled = checked !== undefined;
  const isChecked = isControlled ? checked : innerChecked;

  // Sync with external value changes
  useEffect(() => {
    if (isControlled && checked !== undefined) {
      setInnerChecked(checked);
    }
  }, [isControlled, checked]);

  const handleChange = (newChecked: boolean) => {
    if (!disabled && !loading) {
      if (!isControlled) {
        setInnerChecked(newChecked);
      }
      
      if (onChange) {
        onChange(newChecked);
      }
    }
  };

  return (
    <div className={cx('memori-switch-wrapper', className)}>
      {label && (
        <label htmlFor={id} className="memori-switch--label">
          {label}
        </label>
      )}
      
      <HeadlessSwitch
        checked={isChecked}
        onChange={handleChange}
        className={cx('memori-switch', `memori-switch--${size}`, {
          'memori-switch--checked': isChecked,
          'memori-switch--disabled': disabled || loading,
          'memori-switch--loading': loading
        })}
        disabled={disabled || loading}
        data-testid={testId}
        id={id}
      >
        <span 
          aria-hidden="true" 
          className={cx('memori-switch--handle', {
            'memori-switch--handle-checked': isChecked,
            'memori-switch--handle-disabled': disabled || loading,
            'memori-switch--handle-loading': loading
          })}
        >
          {loading && (
            <span className="memori-switch--loading-icon">
              <svg 
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 24 24"
                fill="none" 
                stroke="currentColor" 
                strokeWidth="2" 
                strokeLinecap="round" 
                strokeLinejoin="round"
                className="memori-switch--loading-spinner"
              >
                <circle cx="12" cy="12" r="10"></circle>
                <path d="M12 6v2"></path>
              </svg>
            </span>
          )}
        </span>
        
        {(checkedChildren || unCheckedChildren) && (
          <span className="memori-switch--inner">
            {isChecked ? checkedChildren : unCheckedChildren}
          </span>
        )}
      </HeadlessSwitch>
    </div>
  );
};

export default Switch;