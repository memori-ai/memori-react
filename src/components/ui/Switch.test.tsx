import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Switch from './Switch';

describe('Switch Component', () => {
  it('renders Switch unchanged', () => {
    const { container } = render(
      <Switch onChange={jest.fn()} />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with label unchanged', () => {
    const { container } = render(
      <Switch
        label="Toggle me"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch checked unchanged', () => {
    const { container } = render(
      <Switch
        checked={true}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with defaultChecked unchanged', () => {
    const { container } = render(
      <Switch
        defaultChecked={true}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch disabled unchanged', () => {
    const { container } = render(
      <Switch
        disabled={true}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch disabled and checked unchanged', () => {
    const { container } = render(
      <Switch
        disabled={true}
        checked={true}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with size small unchanged', () => {
    const { container } = render(
      <Switch
        size="small"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with size large unchanged', () => {
    const { container } = render(
      <Switch
        size="large"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with loading state unchanged', () => {
    const { container } = render(
      <Switch
        loading={true}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with checkedChildren and unCheckedChildren unchanged', () => {
    const { container } = render(
      <Switch
        checkedChildren="ON"
        unCheckedChildren="OFF"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders Switch with data-testid unchanged', () => {
    const { container } = render(
      <Switch
        data-testid="feature-toggle"
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('calls onChange when clicked', () => {
    const handleChange = jest.fn();
    render(<Switch onChange={handleChange} data-testid="test-switch" />);
    
    const switchElement = screen.getByTestId('test-switch');
    fireEvent.click(switchElement);
    
    expect(handleChange).toHaveBeenCalledWith(true);
  });

  it('does not call onChange when disabled', () => {
    const handleChange = jest.fn();
    render(<Switch onChange={handleChange} disabled data-testid="test-switch" />);
    
    const switchElement = screen.getByTestId('test-switch');
    fireEvent.click(switchElement);
    
    expect(handleChange).not.toHaveBeenCalled();
  });

  it('does not call onChange when loading', () => {
    const handleChange = jest.fn();
    render(<Switch onChange={handleChange} loading data-testid="test-switch" />);
    
    const switchElement = screen.getByTestId('test-switch');
    fireEvent.click(switchElement);
    
    expect(handleChange).not.toHaveBeenCalled();
  });
});