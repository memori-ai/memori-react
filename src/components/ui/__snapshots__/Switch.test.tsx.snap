// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Component renders Switch checked unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="true"
      class="memori-switch memori-switch--default memori-switch--checked"
      data-headlessui-state="checked"
      id="headlessui-switch-:r2:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle memori-switch--handle-checked"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch disabled and checked unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="true"
      class="memori-switch memori-switch--default memori-switch--checked memori-switch--disabled"
      data-headlessui-state="checked"
      disabled=""
      id="headlessui-switch-:r5:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle memori-switch--handle-checked memori-switch--handle-disabled"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch disabled unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--default memori-switch--disabled"
      data-headlessui-state=""
      disabled=""
      id="headlessui-switch-:r4:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle memori-switch--handle-disabled"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--default"
      data-headlessui-state=""
      id="headlessui-switch-:r0:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with checkedChildren and unCheckedChildren unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--default"
      data-headlessui-state=""
      id="headlessui-switch-:r9:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle"
      />
      <span
        class="memori-switch--inner"
      >
        OFF
      </span>
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with data-testid unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--default"
      data-headlessui-state=""
      data-testid="feature-toggle"
      id="headlessui-switch-:ra:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with defaultChecked unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="true"
      class="memori-switch memori-switch--default memori-switch--checked"
      data-headlessui-state="checked"
      id="headlessui-switch-:r3:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle memori-switch--handle-checked"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with label unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <label
      class="memori-switch--label"
    >
      Toggle me
    </label>
    <button
      aria-checked="false"
      class="memori-switch memori-switch--default"
      data-headlessui-state=""
      id="headlessui-switch-:r1:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with loading state unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--default memori-switch--disabled memori-switch--loading"
      data-headlessui-state=""
      disabled=""
      id="headlessui-switch-:r8:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle memori-switch--handle-disabled memori-switch--handle-loading"
      >
        <span
          class="memori-switch--loading-icon"
        >
          <svg
            class="memori-switch--loading-spinner"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="M12 6v2"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with size large unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--large"
      data-headlessui-state=""
      id="headlessui-switch-:r7:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle"
      />
    </button>
  </div>
</div>
`;

exports[`Switch Component renders Switch with size small unchanged 1`] = `
<div>
  <div
    class="memori-switch-wrapper"
  >
    <button
      aria-checked="false"
      class="memori-switch memori-switch--small"
      data-headlessui-state=""
      id="headlessui-switch-:r6:"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="memori-switch--handle"
      />
    </button>
  </div>
</div>
`;
